File: c++_code/215.数组中的第k个最大元素.cpp
/*
 * @lc app=leetcode.cn id=215 lang=cpp
 *
 * [215] 数组中的第K个最大元素
 */

// @lc code=start
// 快搜选择
#include <vector>
using namespace std;
class Solution {
public:
    //对一个升序数组来说，第一大的元素就是倒数第一个即n-1,
    //第k大的元素是倒数第k个，即n-k
    int findKthLargest(vector<int>& nums, int k) {
        int left=0,right=nums.size()-1,target=nums.size()-k;
        while(left<right)
        {
            int ans=quick_sort(nums,left,right);
            if(ans==target)
            {
                return nums[ans];
            }
            else if(ans<target)
            {
                left=ans+1;
            }
            else if(ans>target)
            {
                right=ans-1;
            }
        }
        return nums[left];
        
    }

    //快排序数组
    int quick_sort(vector<int>& nums,int l,int r)
    {
        int temp=nums[l];
        while(l<r)
        {
            //从右向左，找到小于target值的那个，
            //把该值放到target值暂存的那个位置。
            while(l<r&&nums[r]>temp) r--;
            nums[l]=nums[r];
            while(l<r&&nums[l]<=temp) l++;
            nums[r]=nums[l];
        }
        nums[l]=temp;
        return l;
    }
};
// @lc code=end

File: js_code/正常顺序/215.数组中的第k个最大元素.js
/*
 * @lc app=leetcode.cn id=215 lang=javascript
 *
 * [215] 数组中的第K个最大元素
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */

//快速选择
/*
快速选择一般用于求解 k-th Element 问题，可以在 O(n) 时间复杂度，O(1) 空间复杂度完成求
解工作。快速选择的实现和快速排序相似，不过只需要找到第 k 大的枢（pivot）即可，不需要对
其左右再进行排序。
*/

var findKthLargest = function(nums, k) {


  //快速选择
  const quick_sort=(nums,l,r)=>{
    let first =l+1,last =r,key=nums[l];
    while(1){
      while(first<r&&nums[first]<=key) first++;
      while(l<last&&nums[last]>=key) last--;
      if(first>=last) break;

      [nums[first], nums[last]]=[nums[last],nums[first]]
    }
    [nums[l], nums[last]]=[nums[last],nums[l]]
    return last

  }

  let l=0,r=nums.length-1,target=nums.length-k
  while(l<r){
    let mid=quick_sort(nums,l,r)
    if(mid===target){
      return nums[mid]
    }else if(mid<target){
      l=mid+1
    }else if(mid>target){
      r=mid-1
    }
  }
  return nums[l]

};
// @lc code=end

